src/main/java/life/qbic/model/download/QbicDataDownloader.java

```diff
@@ -139,140 +163,201 @@ private String getFileName(DataSetFile file) {
-        String fileName = filePath.getFileName().toString();
+        long computedChecksum;
+
+        String fileName = finalPath.getFileName().toString();
         ProgressBar progressBar =
-                new ProgressBar(
-                        fileName, file.getDataSetFile().getFileLength());
+            new ProgressBar(
+                fileName, file.getDataSetFile().getFileLength());
         int bufferSize =
-                (file.getDataSetFile().getFileLength() < defaultBufferSize)
-                        ? (int) file.getDataSetFile().getFileLength()
-                        : defaultBufferSize;
+            (file.getDataSetFile().getFileLength() < defaultBufferSize)
+                ? (int) file.getDataSetFile().getFileLength()
+                : defaultBufferSize;
         byte[] buffer = new byte[bufferSize];
         int bytesRead;
-        LOG.info(String.format("Download of %s is starting", fileName));
-        // read from is to buffer
-        while ((bytesRead = checkedInputStream.read(buffer)) != -1) {
-          progressBar.updateProgress(bufferSize);
-          os.write(buffer, 0, bytesRead);
+        LOG.debug(String.format("Download of %s is starting", fileName));
+        try (InputStream initialStream = file.getInputStream();
+            OutputStream os = Files.newOutputStream(newFile.toPath());
+            CheckedInputStream checkedInputStream = new CheckedInputStream(initialStream,
+                new CRC32())) {
+          // read from is to buffer
+          while ((bytesRead = checkedInputStream.read(buffer)) != -1) {
+            progressBar.updateProgress(bufferSize);
+            os.write(buffer, 0, bytesRead);
+            os.flush();
+          }
+          // flush OutputStream to write any buffered data to file
           os.flush();
+          LOG.debug(String.format("Download of %s has finished", fileName));
+          computedChecksum = checkedInputStream.getChecksum().getValue();
+        } catch (IOException e) {
+          throw new DownloadException(e);
         }
-        initialStream.close();
-        // flush OutputStream to write any buffered data to file
-        os.flush();
-        LOG.info(String.format("Download of %s has finished", fileName));
-        validateChecksum(
-                Long.toHexString(checkedInputStream.getChecksum().getValue()), dataSetFile);
-        if(invalidChecksumOccurred) {
-          notifyUserOfInvalidChecksum(dataSetFile);
-        }
-        os.close();
+        doChecksumLogic(dataSetFile, finalPath, computedChecksum);
+      } else {
+        LOG.info("Skipped empty file " + file.getDataSetFile().getPath());
       }
     }
   }
 
-  private void validateChecksum(String computedChecksumHex, DataSetFile dataSetFile) {
-    String expectedChecksum = Integer.toHexString(dataSetFile.getChecksumCRC32());
+  private void doChecksumLogic(DataSetFile dataSetFile, Path filePath, long computedChecksum) {
```

Comment:
This is mostly used to report the checksum, right? The name was confusing to me without looking what happens.

Reply:
This is used to validate the checksum and to report the result (both in a file and to the log/console)

Reply:
Do you have a suggestion for a method name?

Reply:
verifyAndReportFileIntegrity()
or
checkAndReportFileIntegrity()